File: services/typescript/express/src/routes/handlers/get.ts
import type { Request, RequestHandler, Response } from "express";
import { z } from "zod";
import type { State } from "../../state.js";
import { getLogger } from "../../logger.js";
import { AppErrors, sendError } from "../errors.js";

export function getHandler(state: State): RequestHandler {
  return (req: Request, res: Response) => {
    const logger = getLogger();

    const id = z.uuidv4().safeParse(req.params.id);
    if (!id.success) {
      logger.warn({ method: req.method, path: req.path }, "Invalid path");
      return sendError(res, AppErrors.InvalidPath);
    }

    const task = state.tasks.get(id.data);
    if (!task) {
      logger.warn(
        { id: id.data, method: "GET", path: req.path },
        "Task not found"
      );
      return sendError(res, AppErrors.TaskNotFound);
    }

    logger.info(
      {
        id: id.data,
        secret: task.secret.length,
        operation: task.operation.toString().toLowerCase(),
        method: "GET"
      },
      "Retrieved task"
    );

    return res.status(200).json(task);
  };
}

File: services/python/sanic/src/routes/handlers/get.py
import structlog
from uuid import UUID
from sanic import Request, json
from sanic.response import JSONResponse

from app.state import AppState
from app.task import Task
from routes.errors import send_error, AppErrors


async def get_handler(request: Request, id: UUID) -> JSONResponse:
    logger = structlog.get_logger()

    state: AppState = request.app.ctx.state

    async with request.app.ctx.tasks_lock:
        task: Task | None = state.tasks.get(id)

        if task is None:
            logger.warn(
                "Task not found", id=str(id), method=request.method, path=request.path
            )
            return send_error(AppErrors.TaskNotFound)

    logger.info(
        "Retrieved task",
        id=str(id),
        secret=len(task.secret),
        operation=task.operation.lower(),
        method=request.method,
    )

    return json(task.model_dump())
